File: sdk/python/atlas.py
"""Atlas Python SDK — a zero-dependency client for the Atlas Server.

Atlas issues and verifies offline-verifiable, attenuable delegation tokens
bound to SPIFFE workload identity. This client talks to `cmd/atlas-server`
over HTTP using only the standard library.

    from atlas import AtlasClient, Decision

    atlas = AtlasClient("http://127.0.0.1:8087", api_key=None)

    grant = atlas.issue(
        principal="spiffe://domain-a.test/workload/payments-api",
        delegate="spiffe://domain-b.test/agent/booking-worker",
        scope=["read:orders", "write:audit"],
        ttl_seconds=3600,
    )
    result = atlas.verify(grant.record)
    assert result.decision == Decision.ACCEPT

    atlas.revoke(grant.instance)
    assert atlas.verify(grant.record).decision == Decision.REJECT
"""
from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class Decision(str, Enum):
    ACCEPT = "accept"
    REJECT = "reject"
    INCONCLUSIVE = "inconclusive"


class AtlasError(Exception):
    """Raised when the server returns an error or is unreachable."""


@dataclass
class Grant:
    record: str
    instance: str
    principal: str
    delegate: str
    scope: list[str]
    expires_at: str

    @classmethod
    def _from(cls, d: dict[str, Any]) -> "Grant":
        return cls(
            record=d["record"], instance=d["instance"],
            principal=d.get("principal", ""), delegate=d.get("delegate", ""),
            scope=d.get("scope", []), expires_at=d.get("expiresAt", ""),
        )


@dataclass
class TraceEntry:
    check: str
    outcome: str
    cause: str


@dataclass
class VerifyResult:
    decision: Decision
    accept: bool
    causes: list[str]
    latency_micros: int
    trace: list[TraceEntry] = field(default_factory=list)

    @classmethod
    def _from(cls, d: dict[str, Any]) -> "VerifyResult":
        return cls(
            decision=Decision(d["decision"]) if d["decision"] in Decision._value2member_map_ else d["decision"],
            accept=bool(d.get("accept")),
            causes=d.get("causes") or [],
            latency_micros=int(d.get("latencyMicros", 0)),
            trace=[TraceEntry(t.get("check", ""), t.get("outcome", ""), t.get("cause", "")) for t in d.get("trace") or []],
        )


class AtlasClient:
    """A thin HTTP client for the Atlas Server."""

    def __init__(self, base_url: str = "http://127.0.0.1:8087",
                 api_key: str | None = None, timeout: float = 8.0) -> None:
        self.base = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    # -- core --

    def _call(self, method: str, path: str, body: dict | None = None) -> Any:
        data = json.dumps(body).encode() if body is not None else None
        req = urllib.request.Request(self.base + path, data=data, method=method)
        req.add_header("Content-Type", "application/json")
        if self.api_key:
            req.add_header("Authorization", "Bearer " + self.api_key)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read()
        except urllib.error.HTTPError as e:
            raw = e.read()
            try:
                msg = json.loads(raw).get("error", raw.decode())
            except Exception:
                msg = raw.decode(errors="replace")
            raise AtlasError(f"{e.code}: {msg}") from None
        except urllib.error.URLError as e:
            raise AtlasError(f"atlas-server unreachable at {self.base} ({e.reason})") from None
        return json.loads(raw) if raw else {}

    # -- operations --

    def health(self) -> bool:
        try:
            return self._call("GET", "/health").get("status") == "ok"
        except AtlasError:
            return False

    def version(self) -> dict:
        return self._call("GET", "/version")

    def issue(self, principal: str, delegate: str, scope: list[str],
              ttl_seconds: int = 3600) -> Grant:
        return Grant._from(self._call("POST", "/issue", {
            "principal": principal, "delegate": delegate,
            "scope": scope, "ttlSeconds": ttl_seconds,
        }))

    def verify(self, record: str) -> VerifyResult:
        return VerifyResult._from(self._call("POST", "/verify", {"record": record}))

    def revoke(self, instance: str) -> None:
        self._call("POST", "/revoke", {"instance": instance})

    def delegations(self) -> list[dict]:
        return self._call("GET", "/delegations").get("delegations", [])

    def audit(self, limit: int = 50) -> list[dict]:
        return self._call("GET", f"/audit?limit={limit}").get("events", [])

    def graph(self) -> dict:
        return self._call("GET", "/graph")

    def stats(self) -> dict:
        return self._call("GET", "/stats")


__all__ = ["AtlasClient", "Grant", "VerifyResult", "TraceEntry", "Decision", "AtlasError"]

File: sdk/typescript/atlas.ts
/**
 * Atlas TypeScript SDK — a zero-dependency client for the Atlas Server.
 *
 * Atlas issues and verifies offline-verifiable, attenuable delegation tokens
 * bound to SPIFFE workload identity. This client talks to `cmd/atlas-server`
 * over HTTP using only the platform `fetch` (Node 18+, Deno, Bun, browsers).
 *
 * ```ts
 * import { AtlasClient } from "./atlas.ts";
 *
 * const atlas = new AtlasClient("http://127.0.0.1:8087");
 * const grant = await atlas.issue({
 *   principal: "spiffe://domain-a.test/workload/payments-api",
 *   delegate:  "spiffe://domain-b.test/agent/booking-worker",
 *   scope:     ["read:orders", "write:audit"],
 *   ttlSeconds: 3600,
 * });
 * const result = await atlas.verify(grant.record);
 * console.log(result.decision); // "accept"
 *
 * await atlas.revoke(grant.instance);
 * (await atlas.verify(grant.record)).decision; // "reject"
 * ```
 *
 * The API surface mirrors the Python SDK (`sdk/python/atlas.py`) one-for-one.
 */

/** A verification verdict. Deterministic: reconstructable from the record + trace. */
export type Decision = "accept" | "reject" | "inconclusive";
export const Decision = {
  Accept: "accept",
  Reject: "reject",
  Inconclusive: "inconclusive",
} as const;

/** Raised when the server returns an error or is unreachable. */
export class AtlasError extends Error {
  status: number;
  constructor(message: string, status = 0) {
    super(message);
    this.name = "AtlasError";
    this.status = status;
  }
}

export interface Grant {
  record: string;
  instance: string;
  principal: string;
  delegate: string;
  scope: string[];
  issuedAt: string;
  expiresAt: string;
}

export interface TraceEntry {
  check: string;
  outcome: string;
  cause: string;
  detail?: string;
}

export interface VerifyResult {
  decision: Decision;
  accept: boolean;
  causes: string[];
  latencyMicros: number;
  trace: TraceEntry[];
}

export interface IssueParams {
  principal: string;
  delegate: string;
  scope: string[];
  ttlSeconds?: number;
}

export interface AtlasClientOptions {
  apiKey?: string;
  /** Request timeout in milliseconds (default 8000). */
  timeoutMs?: number;
}

/** A thin, promise-based HTTP client for the Atlas Server. */
export class AtlasClient {
  private base: string;
  private apiKey: string | undefined;
  private timeoutMs: number;

  constructor(baseUrl = "http://127.0.0.1:8087", opts: AtlasClientOptions = {}) {
    this.base = baseUrl.replace(/\/+$/, "");
    this.apiKey = opts.apiKey;
    this.timeoutMs = opts.timeoutMs ?? 8000;
  }

  // -- core --

  private async call<T>(method: string, path: string, body?: unknown): Promise<T> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) headers["Authorization"] = "Bearer " + this.apiKey;

    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
    let resp: Response;
    try {
      resp = await fetch(this.base + path, {
        method,
        headers,
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: ctrl.signal,
      });
    } catch (e) {
      const reason = e instanceof Error ? e.message : String(e);
      throw new AtlasError(`atlas-server unreachable at ${this.base} (${reason})`);
    } finally {
      clearTimeout(timer);
    }

    const raw = await resp.text();
    if (!resp.ok) {
      let msg = raw;
      try {
        msg = (JSON.parse(raw) as { error?: string }).error ?? raw;
      } catch {
        /* non-JSON body: use raw text */
      }
      throw new AtlasError(`${resp.status}: ${msg}`, resp.status);
    }
    return (raw ? JSON.parse(raw) : {}) as T;
  }

  // -- operations (mirror sdk/python/atlas.py) --

  /** True when the server is reachable and healthy; never throws. */
  async health(): Promise<boolean> {
    try {
      const r = await this.call<{ status?: string }>("GET", "/health");
      return r.status === "ok";
    } catch {
      return false;
    }
  }

  /** Orchestrator readiness: true only when a fresh revocation snapshot is held. */
  async ready(): Promise<boolean> {
    try {
      const r = await this.call<{ ready?: boolean }>("GET", "/readyz");
      return r.ready === true;
    } catch {
      return false;
    }
  }

  version(): Promise<Record<string, unknown>> {
    return this.call("GET", "/version");
  }

  issue(params: IssueParams): Promise<Grant> {
    return this.call<Grant>("POST", "/issue", {
      principal: params.principal,
      delegate: params.delegate,
      scope: params.scope,
      ttlSeconds: params.ttlSeconds ?? 3600,
    });
  }

  verify(record: string): Promise<VerifyResult> {
    return this.call<VerifyResult>("POST", "/verify", { record });
  }

  async revoke(instance: string): Promise<void> {
    await this.call("POST", "/revoke", { instance });
  }

  async delegations(): Promise<Record<string, unknown>[]> {
    const r = await this.call<{ delegations?: Record<string, unknown>[] }>("GET", "/delegations");
    return r.delegations ?? [];
  }

  async audit(limit = 50): Promise<Record<string, unknown>[]> {
    const r = await this.call<{ events?: Record<string, unknown>[] }>("GET", `/audit?limit=${limit}`);
    return r.events ?? [];
  }

  graph(): Promise<Record<string, unknown>> {
    return this.call("GET", "/graph");
  }

  stats(): Promise<Record<string, unknown>> {
    return this.call("GET", "/stats");
  }
}
